VultisigApp/VultisigApp/Chains/KyberSwaps.swift

```diff
@@ -0,0 +1,122 @@
+                }
+            }
+            $0.toAddress = keysignPayload.coin.contractAddress
+        }
+        let inputData = try EVMHelper.getHelper(coin: keysignPayload.coin).getPreSignedInputData(
+            signingInput: approveInput,
+            keysignPayload: keysignPayload
+        )
+        return inputData
+    }
+
+    func getPreSignedApproveImageHash(approvePayload: ERC20ApprovePayload, keysignPayload: KeysignPayload) throws -> [String] {
+        let inputData = try getPreSignedApproveInputData(
+            approvePayload: approvePayload,
+            keysignPayload: keysignPayload
+        )
+        let hashes = TransactionCompiler.preImageHashes(coinType: keysignPayload.coin.coinType, txInputData: inputData)
+        let preSigningOutput = try TxCompilerPreSigningOutput(serializedBytes: hashes)
+        return [preSigningOutput.dataHash.hexString]
+    }
+
+    func getSignedApproveTransaction(approvePayload: ERC20ApprovePayload, keysignPayload: KeysignPayload, signatures: [String: TssKeysignResponse]) throws -> SignedTransactionResult {
+        let inputData = try getPreSignedApproveInputData(
+            approvePayload: approvePayload,
+            keysignPayload: keysignPayload
+        )
+        let signedEvmTx = try EVMHelper.getHelper(coin: keysignPayload.coin).getSignedTransaction(vaultHexPubKey: vaultHexPublicKey, vaultHexChainCode: vaultHexChainCode, inputData: inputData, signatures: signatures)
+        return signedEvmTx
+    }
+}
+
+private extension KyberSwaps {
+
+    func getPreSignedInputData(quote: KyberSwapQuote, keysignPayload: KeysignPayload, incrementNonce: Bool) throws -> Data {
+        let input = EthereumSigningInput.with {
+            $0.toAddress = quote.tx.to
+            $0.transaction = .with {
+                $0.contractGeneric = .with {
+                    $0.amount = (BigUInt(quote.tx.value) ?? BigUInt.zero).serialize()
+                    $0.data = Data(hex: quote.tx.data.stripHexPrefix())
+                }
+            }
+        }
+
+        let gasPrice = BigUInt(quote.tx.gasPrice) ?? BigUInt("20000000000") // Use gasPrice from quote or 20 Gwei default
+        
+        // Apply chain-specific gas buffer based on chain costs and execution characteristics
+        let baseGas = Int64(quote.data.gas) ?? 600000 // Use raw API gas, not pre-buffered quote.tx.gas
+        let gasMultiplierTimes10: Int64
+        
+        switch keysignPayload.coin.chain {
+        case .ethereum:
+            gasMultiplierTimes10 = 14 // 40% buffer - conservative for expensive Ethereum gas
+        case .arbitrum, .optimism, .base, .polygon, .avalanche, .bscChain:
+            gasMultiplierTimes10 = 20 // 100% buffer - L2s have cheap gas and complex routing
+        default:
+            gasMultiplierTimes10 = 16 // 60% buffer - reasonable default for other chains
+        }
+        
+        let bufferedGas = (baseGas * gasMultiplierTimes10) / 10
+        let gas = BigUInt(bufferedGas)
```

Comment:
Can you please move these logic that update gas to the place where we create quote?  Make sure the gas & gas Price is correct in `quote` 

In this way , the pair doesn't need to have this logic , and make easier for cross device keysign